Engine/CommandInfoCache.cs

```diff
@@ -0,0 +1,119 @@
+        private readonly ConcurrentDictionary<CommandLookupKey, Lazy<CommandInfo>> _commandInfoCache;
+
+        private readonly Helper _helperInstance;
+
+        /// <summary>
+        /// Create a fresh command info cache instance.
+        /// </summary>
+        public CommandInfoCache(Helper pssaHelperInstance)
+        {
+            _commandInfoCache = new ConcurrentDictionary<CommandLookupKey, Lazy<CommandInfo>>();
+            _helperInstance = pssaHelperInstance;
+        }
+
+        /// <summary>
+        /// Retrieve a command info object about a command.
+        /// </summary>
+        /// <param name="commandName">Name of the command to get a commandinfo object for.</param>
+        /// <param name="aliasName">The alias of the command to be used in the cache key. If not given, uses the command name.</param>
+        /// <param name="commandTypes">What types of command are needed. If omitted, all types are retrieved.</param>
+        /// <returns></returns>
+        public CommandInfo GetCommandInfo(string commandName, string aliasName = null, CommandTypes? commandTypes = null)
+        {
+            if (string.IsNullOrWhiteSpace(commandName))
+            {
+                return null;
+            }
+
+            // If alias name is given, we store the entry under that, but search with the command name
+            var key = new CommandLookupKey(aliasName ?? commandName, commandTypes);
+
+            // Atomically either use PowerShell to query a command info object, or fetch it from the catch
+            return _commandInfoCache.GetOrAdd(key, new Lazy<CommandInfo>(() => GetCommandInfoInternal(commandName, commandTypes))).Value;
+        }
+
+        /// <summary>
+        /// Retrieve a command info object about a command.
+        /// </summary>
+        /// <param name="commandName">Name of the command to get a commandinfo object for.</param>
+        /// <param name="commandTypes">What types of command are needed. If omitted, all types are retrieved.</param>
+        /// <returns></returns>
+        [Obsolete("Alias lookup is expensive and should not be relied upon for command lookup")]
+        public CommandInfo GetCommandInfoLegacy(string commandOrAliasName, CommandTypes? commandTypes = null)
+        {
+            string commandName = _helperInstance.GetCmdletNameFromAlias(commandOrAliasName);
+
+            return string.IsNullOrEmpty(commandName)
+                ? GetCommandInfo(commandOrAliasName, commandTypes: commandTypes)
+                : GetCommandInfo(commandName, aliasName: commandOrAliasName, commandTypes: commandTypes);
+        }
+
+        /// <summary>
+        /// Get a CommandInfo object of the given command name
+        /// </summary>
+        /// <returns>Returns null if command does not exists</returns>
+        private static CommandInfo GetCommandInfoInternal(string cmdName, CommandTypes? commandType)
+        {
+            using (var ps = System.Management.Automation.PowerShell.Create())
```

Comment:
There is a chance to further improve the perf by using a static RunspacePool, instead of creating/disposing a Runspace for every call to this method.
Of course, you will need to decide the size of the pool carefully based on the typical scenarios.

The code would be something like this:
```
staitc RunspacePool rps = RunspaceFactory.CreateRunspacePool(minRunspaces: 1, maxRunspaces: 5)
...
private static CommandInfo GetCommandInfoInternal(string cmdName, CommandTypes? commandType)
{
    using (var ps = System.Management.Automation.PowerShell.Create())
    {
        ps.RunspacePool = rps;
        ps.AddCommand("Get-Command").Invoke()
        ...
    }
}
```

Reply:
Yes, I have tried something like this already in the past but could not get any significant measurable differences in terms of time. I think because all rules are run in separate threads, the rule that takes the longest is the weakest link in the chain and therefore the bottleneck.

Reply:
Using a `RunspacePool` here is mainly for two purpose:
1. Make this method run faster
    - `Runspace.Open` is expensive.
    - module auto-loading and the cache populated during command discovery can be reused
    - `RunspacePool` may cause some extent of synchronization among the threads (more threads than the number of `Runspaces` get in this method). But only one invocation will be done by each thread, so the introduced synchronization wouldn't be that bad.
2. Reduce the GC pressure
    - Open and dispose a `Runspace` in every thread that gets in this method would produce a lot shot living objects.

As for how effective it is to the overall run time improvement, maybe not very -- measurement speaks the truth :) But it looks to me a low hanging fruit.